src/Microsoft.Data.SqlClient/src/Microsoft/Data/ProviderBase/DbReferenceCollection.cs

```diff
@@ -0,0 +1,280 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+
+namespace Microsoft.Data.ProviderBase
+{
+    internal abstract class DbReferenceCollection
+    {
+        #region Constants
+        // Time to wait (in ms) between attempting to get the _itemLock
+        private const int LockPollTime = 100;
+
+        // Default size for the collection, and the amount to grow every time the collection is full
+        private const int DefaultCollectionSize = 20;
+        #endregion
+
+        #region Fields
+        // The collection of items we are keeping track of
+        private CollectionEntry[] _items;
+
+        // Used to synchronize access to the _items collection
+        private readonly object _itemLock;
+
+        // (#ItemsAdded - #ItemsRemoved) - This estimates the number of items that we should have
+        // (but doesn't take into account item targets being GC'd)
+        private int _estimatedCount;
+
+        // Location of the last item in _items
+        private int _lastItemIndex;
+
+        // Indicates that the collection is currently being notified (and, therefore, about to be cleared)
+        private volatile bool _isNotifying;
+        #endregion
+
+        private struct CollectionEntry
+        {
+            private int _refInfo;              // information about the reference
+            private WeakReference<object> _weakReference;   // the reference itself.
+
+            public void SetTNewTarget(int refInfo, object target)
```

Comment:
Typo! I'd rather `AddTarget` instead of `SetNewTarget`.
```suggestion
            public void SetNewTarget(int refInfo, object target)
```